### PR TITLE
(3.4) dnn: fix API - explicit ctors, const methods

### PR DESCRIPTION
**Pull request by alalek**
Validate:
- [x] samples/dnn/dasiamrpn_tracker.py
- [x] Halide build configuration
- [x] OpenVINO build configuration
- [x] OpenVINO build configuration (NN builder - 2019r3.0)

<cut/>

```
force_builders=Custom,Custom Win,Custom Mac

buildworker:Docs=linux-4,linux-6
build_image:Docs=docs-js:18.04


build_image:Custom=halide:16.04
Xbuild_image:Custom=ubuntu-openvino-2019r3.0:16.04
Xbuild_image:Custom=ubuntu-openvino-2021.4.0:20.04
Xbuild_image:Custom=ubuntu-openvino-2021.3.0:20.04
Xbuild_image:Custom=ubuntu-openvino-2021.2.0:20.04
Xbuild_image:Custom=ubuntu-openvino-2020.3.0:16.04
build_image:Custom Win=openvino-2021.4.2
Xbuild_image:Custom Win=openvino-2021.3.0-vc16
build_image:Custom Mac=openvino-2021.4.2

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-6
Xbuildworker:Custom=linux-5
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*

buildworker:Custom Win=windows-3
test_opencl:Custom Win=ON
test_bigdata:Custom Win=1
test_filter:Custom Win=*
```

### CODE REVIEW

**Review by asmorkalov — APPROVED** [review]

File: modules/dnn/src/dnn.cpp [hunk 1]
```diff
@@ -1394,23 +1393,23 @@ struct Net::Impl : public detail::NetImplBase
...
     {
-        MapIdToLayerData::iterator it = layers.find(id);
+        MapIdToLayerData::const_iterator it = layers.find(id);
 
         if (it == layers.end())
             CV_Error(Error::StsObjectNotFound, format("Layer with requested id=%d not found", id));
 
-        return it->second;
+        return const_cast<LayerData&>(it->second);
     }
```
Comment (rogday) [inline modules/dnn/src/dnn.cpp:1410]: We are returning non-const reference from const method. Why do we need that?

File: modules/dnn/src/dnn.cpp [hunk 2]
```diff
@@ -1464,13 +1463,13 @@ struct Net::Impl : public detail::NetImplBase
...
 
-    std::vector<LayerPin> getLayerOutPins(const String &layerName)
+    std::vector<LayerPin> getLayerOutPins(const String &layerName) const
     {
         int lid = (layerName.empty()) ? 0 : getLayerId(layerName);
 
         std::vector<LayerPin> pins;
 
-        for (int i = 0; i < layers[lid].outputBlobs.size(); i++)
+        for (int i = 0; i < const_cast<MapIdToLayerData&>(layers)[lid].outputBlobs.size(); i++)
```
Comment (rogday) [inline modules/dnn/src/dnn.cpp:1472]: I think, we either should check that lid exists in layers, or mark layers as mutable, or don't mark this method as const. I think, the use of const_cast is unjustified here.

File: modules/dnn/src/dnn.cpp [hunk 3]
```diff
@@ -3187,19 +3180,19 @@ struct Net::Impl : public detail::NetImplBase
...
+    Mat getBlob(const LayerPin& pin) const
     {
         CV_TRACE_FUNCTION();
 
         if (!pin.valid())
             CV_Error(Error::StsObjectNotFound, "Requested blob not found");
 
-        LayerData &ld = layers[pin.lid];
+        const LayerData &ld = const_cast<MapIdToLayerData&>(layers)[pin.lid];
         if ((size_t)pin.oid >= ld.outputBlobs.size())
```
Comment (rogday) [inline modules/dnn/src/dnn.cpp:3196]: If we create `[pin.lid]` in layers in previous line, its `outputBlobs.size()` is zero and the condition is true, so we can get rid of const_cast.

**Review by rogday — CHANGES_REQUESTED** [review]
Thank you for contribution!

File: modules/dnn/src/dnn.cpp (hunk 1 above)
Comment (alalek) [reply modules/dnn/src/dnn.cpp:1410]: This change requires too much changes due to current.
Unfortunately `LayerData` is not not preserved unmodified. It is part of "graph" model and manages graph connections too.

File: modules/dnn/src/dnn.cpp (hunk 2 above)
Comment (alalek) [reply modules/dnn/src/dnn.cpp]: Updated

File: modules/dnn/src/dnn.cpp (hunk 3 above)
Comment (alalek) [reply modules/dnn/src/dnn.cpp]: Updated

**Comment by alalek** [comment]
:+1:

**Pull request merged**